### ISSUE TITLE
Cannot build libcrypto library on Windows with multithreading support but without thread pool

### ISSUE THREAD

**Issue by manison** [openssl/openssl#25337]
I'm trying to build a libcrypto library for Windows with lower footprint. I want to keep multithreading support (locking) but in my application I don't need a thread pool support. Configuration is as follows:

```
perl Configure VC-WIN64A-HYBRIDCRT no-dh no-engine no-http no-ocsp no-async no-thread-pool no-quic
```
[config.txt](https://github.com/user-attachments/files/16818388/config.txt)

However with this configuration the build ends with the following error:

```
libcrypto-shlib-threads_win.obj : error LNK2019: unresolved external symbol ossl_crypto_mutex_new referenced in function ossl_rcu_lock_new
libcrypto-shlib-threads_win.obj : error LNK2019: unresolved external symbol ossl_crypto_mutex_lock referenced in function ossl_rcu_write_lock
libcrypto-shlib-threads_win.obj : error LNK2019: unresolved external symbol ossl_crypto_mutex_unlock referenced in function ossl_rcu_write_unlock
libcrypto-shlib-threads_win.obj : error LNK2019: unresolved external symbol ossl_crypto_mutex_free referenced in function ossl_rcu_lock_free
libcrypto-shlib-threads_win.obj : error LNK2019: unresolved external symbol ossl_crypto_condvar_new referenced in function ossl_rcu_lock_new
libcrypto-shlib-threads_win.obj : error LNK2019: unresolved external symbol ossl_crypto_condvar_wait referenced in function ossl_synchronize_rcu
libcrypto-shlib-threads_win.obj : error LNK2019: unresolved external symbol ossl_crypto_condvar_broadcast referenced in function ossl_synchronize_rcu
libcrypto-shlib-threads_win.obj : error LNK2019: unresolved external symbol ossl_crypto_condvar_free referenced in function ossl_rcu_lock_free
libcrypto-3-x64.dll : fatal error LNK1120: 8 unresolved externals
```

This is because unconditionally compiled RCU functions

https://github.com/openssl/openssl/blob/0cd9dd703ea575699b2d3cd74f1b8224447f4352/crypto/threads_win.c#L158-L182

depend on conditionally compiled mutex/condvar functions

https://github.com/openssl/openssl/blob/0cd9dd703ea575699b2d3cd74f1b8224447f4352/crypto/thread/arch/thread_win.c#L97-L105

I believe the POSIX implementation suffers with the same issue.

Probably the arch source code should be either compiled unconditionally or (maybe better) the mutex/condvar functions should be moved to separate files and compiled based on different condition.

**Comment by nhorman** [comment]
I agree, the inclusion of the ossl_crypto_mutex symbols is conditional on both thread-pool and quic being enabled, which doesn't make much sense to me.

It also builds thread_[win|posix|none] unilaterally and relies on defines in each source file to be empty at appropriate times, that seems inefficient (though its in keeping with how we do that in crypto/).  let me try work a patch for you

**Comment by nhorman** [comment]
Can you try with this patch:
```
diff --git a/crypto/thread/build.info b/crypto/thread/build.info
index 191e25e20d..f1eddc5bad 100644
--- a/crypto/thread/build.info
+++ b/crypto/thread/build.info
@@ -4,18 +4,9 @@ $THREADS_ARCH=\
       arch.c  \
       arch/thread_win.c arch/thread_posix.c arch/thread_none.c
 
-IF[{- !$disabled{'thread-pool'} -}]
-  IF[{- !$disabled{quic} -}]
-    SHARED_SOURCE[../../libssl]=$THREADS_ARCH
-  ENDIF
-  $THREADS=\
-        api.c internal.c $THREADS_ARCH
-ELSE
-  IF[{- !$disabled{quic} -}]
-    SOURCE[../../libssl]=$THREADS_ARCH
-  ENDIF
-  $THREADS=api.c
-ENDIF
+$THREADS=\
+      api.c internal.c $THREADS_ARCH
 
-SOURCE[../../libcrypto]=$THREADS
-SOURCE[../../providers/libfips.a]=$THREADS
+SOURCE[../../libssl]=api.c internal.c $THREADS_ARCH
+SOURCE[../../libcrypto]=api.c internal.c $THREADS_ARCH
+SOURCE[../../providers/libfips.a]=api.c internal.c $THREADS_ARCH
```

Its pretty heavy handed, as it just builds everything unilaterally now, but thats probably needed without a larger refactoring

**Comment by manison** [comment]
Wow, that was real quick. I'm out of office now, will try on Monday and let you know. Thanks.

**Comment by manison** [comment]
I can confirm the patch fixes the issue, thanks.

**Comment by t8m** [comment]
I believe we should do a proper fix here, i.e. move the mutex stuff from thread pool files into the base thread locking support files.

**Comment by nhorman** [comment]
yeah, that might make more sense @t8m. Let me give that a shot

**Comment by nhorman** [comment]
wait a second, doing has some implications I didn't consider.  the threads library that was created in crytpo_thread builds into its own library that is independently pulled into libcrypto/libssl/libfips.a.  If we merge the code to threads_posix/threads_win/threads_none, it becomes part of the UTIL_COMMON object set that is only pulled into libcrypto and libfips.a (not libssl), add that last bit implies that we pull in a whole bunch more code that we didn't previously.  We would need to split out the threads libraries and include them independently to avoid that.  Is that something we really want to do?

**Comment by t8m** [comment]
> wait a second, doing has some implications I didn't consider. the threads library that was created in crytpo_thread builds into its own library that is independently pulled into libcrypto/libssl/libfips.a. If we merge the code to threads_posix/threads_win/threads_none, it becomes part of the UTIL_COMMON object set that is only pulled into libcrypto and libfips.a (not libssl), add that last bit implies that we pull in a whole bunch more code that we didn't previously. We would need to split out the threads libraries and include them independently to avoid that. Is that something we really want to do?

Hmm... then we need to have 3 sets of functions:
 1. the old rwlocks, etc. that are pulled into libcrypto and libfips
 2. the mutexes and rcu locks that are pulled into libcrypto, libfips and libssl
 3. the thread pool stuff that is pulled into libcrypto and libssl and is (at least partially) not built when no-thread-pool is specified

**Comment by nhorman** [comment]
That feels like a larger refactoring.   My feeling would be we should take this patch to fix the configuration issue now, and open a new issue to address the larger refactor.  Let me know if you agree, and I'll open the subsequent issue.
